notebooks/_06-Architecture.md

```diff
@@ -0,0 +1,429 @@
+---
+jupyter:
+  jupytext:
+    formats: ipynb,md
+    text_representation:
+      extension: .md
+      format_name: markdown
+      format_version: '1.2'
+      jupytext_version: 1.3.3
+  kernelspec:
+    display_name: Python 3
+    language: python
+    name: python3
+---
+
+# Architecture
+
+
+## 1. At a “high” level, how will the project deliver each of the needed requirements?
+- **Try to think in terms of some manageable (e.g. 10-20) steps**
+- **List needed intermediate products (e.g. pseudo-code, unit tests, etc.)**
+- **Try to be concise but comprehensive?**
+
+The project is split into two tasks:
+
+The first task will be based on simulation work and will go as follows:
+
+For a causal graph where all Xs are independent:
+ - Ingest data
+ - Manipulate data as needed, the output of this step will be a cleaned dataset
+ - Specify a choice model based on the data at hand, the specification should follow required conventions by third party dependencies
+ - The cleaned dataset will be used in a function that simulates choice data based on the specified model, resulting in a dataset with simulated choices
+ - The dataset will be used to estimate a choice model and recover the necessary parameters. These parameters can be stored in a dictionary.
+ - The previous two steps will be repeated for N times (say 100) and the parameters from each iteration will be stored in a dictionary entry.
+ - A function that takes in the parameters of the estimated model and results in the percentage of time that the true parameter is recovered.
```

Comment:
How will we deal with the fact that the parameter vector is multivariate?

Will we report parameter recovery percentages separately for each parameter?

Will we have some way of aggregating results over all parameters and then reporting only a single percentage?

Or will we report parameter recovery percentages for a subset of parameters? (perhaps just the parameter(s) we expect to not be recovered well?).

Reply:
That sounds good to me--track each parameter individually, and we can do whatever afterwards.

An easy way to do this might be something along the lines of 
```
simulated_parameters = np.empty((NUM_SIMULATIONS, num_choice_model_parameters), dtype=float)
for i in range(NUM_SIMULATIONS):
    current_data = simulate_data()
    simulated_parameters[i, :] = estimate_parameters(current_data)
np.savetxt(SIM_EXPERIMENT_RESULT_PATH, simulated_paramters, fmt='%.18f')
```

Reply:
ah I see, if you look at the code I have now, all it does is stores each of the fitted model objects from pylogit as a dictionary value. We can access every single instance of that object by calling it individually. But I like this as well!!

Reply:
Yes what you did makes sense and completely works!